File: typescript/src/algorithm.ts
import * as crypto from "node:crypto";
import type { Algorithm } from "./types.js";
import { UnknownAlgorithmError, InvalidKeyError } from "./errors.js";

/** Sign data with the given algorithm and key material. */
export async function algorithmSign(
  alg: Algorithm,
  key: crypto.KeyObject,
  data: Uint8Array,
): Promise<Uint8Array> {
  switch (alg) {
    case "rsa-pss-sha256":
      return signRSAPSS(key, data, "sha256", 32);
    case "rsa-pss-sha384":
      return signRSAPSS(key, data, "sha384", 48);
    case "rsa-pss-sha512":
      return signRSAPSS(key, data, "sha512", 64);
    case "rsa-v1_5-sha256":
      return signRSAv15(key, data);
    case "ecdsa-p256-sha256":
      return signECDSA(key, data, "sha256", 64);
    case "ecdsa-p384-sha384":
      return signECDSA(key, data, "sha384", 96);
    case "ecdsa-p521-sha512":
      return signECDSA(key, data, "sha512", 132);
    case "ed25519":
      return signEd25519(key, data);
    default:
      throw new UnknownAlgorithmError(alg);
  }
}

/** Verify data with the given algorithm and key material. */
export async function algorithmVerify(
  alg: Algorithm,
  key: crypto.KeyObject,
  data: Uint8Array,
  signature: Uint8Array,
): Promise<boolean> {
  switch (alg) {
    case "rsa-pss-sha256":
      return verifyRSAPSS(key, data, signature, "sha256", 32);
    case "rsa-pss-sha384":
      return verifyRSAPSS(key, data, signature, "sha384", 48);
    case "rsa-pss-sha512":
      return verifyRSAPSS(key, data, signature, "sha512", 64);
    case "rsa-v1_5-sha256":
      return verifyRSAv15(key, data, signature);
    case "ecdsa-p256-sha256":
      return verifyECDSA(key, data, signature, "sha256", 64);
    case "ecdsa-p384-sha384":
      return verifyECDSA(key, data, signature, "sha384", 96);
    case "ecdsa-p521-sha512":
      return verifyECDSA(key, data, signature, "sha512", 132);
    case "ed25519":
      return verifyEd25519(key, data, signature);
    default:
      throw new UnknownAlgorithmError(alg);
  }
}

/** Sign with HMAC using the specified hash algorithm. */
export async function signHMAC(
  secret: Uint8Array,
  data: Uint8Array,
  hash: "sha256" | "sha384" | "sha512" = "sha256",
): Promise<Uint8Array> {
  const hmac = crypto.createHmac(hash, secret);
  hmac.update(data);
  return new Uint8Array(hmac.digest());
}

/** Verify with HMAC using the specified hash algorithm. */
export async function verifyHMAC(
  secret: Uint8Array,
  data: Uint8Array,
  signature: Uint8Array,
  hash: "sha256" | "sha384" | "sha512" = "sha256",
): Promise<boolean> {
  const expected = await signHMAC(secret, data, hash);
  if (expected.length !== signature.length) return false;
  return crypto.timingSafeEqual(expected, signature);
}

// --- RSA-PSS ---

function signRSAPSS(
  key: crypto.KeyObject,
  data: Uint8Array,
  hash: string,
  saltLength: number,
): Uint8Array {
  if (key.type !== "private") {
    throw new InvalidKeyError("expected private key for RSA-PSS signing");
  }
  return new Uint8Array(
    crypto.sign(hash, data, {
      key,
      padding: crypto.constants.RSA_PKCS1_PSS_PADDING,
      saltLength,
    }),
  );
}

function verifyRSAPSS(
  key: crypto.KeyObject,
  data: Uint8Array,
  signature: Uint8Array,
  hash: string,
  saltLength: number,
): boolean {
  if (key.type !== "public") {
    throw new InvalidKeyError("expected public key for RSA-PSS verification");
  }
  return crypto.verify(
    hash,
    data,
    {
      key,
      padding: crypto.constants.RSA_PKCS1_PSS_PADDING,
      saltLength,
    },
    signature,
  );
}

// --- RSA PKCS1 v1.5 ---

function signRSAv15(key: crypto.KeyObject, data: Uint8Array): Uint8Array {
  if (key.type !== "private") {
    throw new InvalidKeyError("expected private key for RSA PKCS1v1.5 signing");
  }
  return new Uint8Array(
    crypto.sign("sha256", data, {
      key,
      padding: crypto.constants.RSA_PKCS1_PADDING,
    }),
  );
}

function verifyRSAv15(
  key: crypto.KeyObject,
  data: Uint8Array,
  signature: Uint8Array,
): boolean {
  if (key.type !== "public") {
    throw new InvalidKeyError("expected public key for RSA PKCS1v1.5 verification");
  }
  return crypto.verify(
    "sha256",
    data,
    {
      key,
      padding: crypto.constants.RSA_PKCS1_PADDING,
    },
    signature,
  );
}

// --- ECDSA ---
// RFC 9421 Section 3.3.3: r || s in raw big-endian format
// P-256: 32+32 = 64 bytes, P-384: 48+48 = 96 bytes, P-521: 66+66 = 132 bytes

function signECDSA(
  key: crypto.KeyObject,
  data: Uint8Array,
  hash: string,
  expectedLength: number,
): Uint8Array {
  if (key.type !== "private") {
    throw new InvalidKeyError("expected private key for ECDSA signing");
  }
  const sig = crypto.sign(hash, data, { key, dsaEncoding: "ieee-p1363" });
  const result = new Uint8Array(sig);
  if (result.length !== expectedLength) {
    throw new InvalidKeyError(
      `ECDSA signature length mismatch: expected ${expectedLength}, got ${result.length}`,
    );
  }
  return result;
}

function verifyECDSA(
  key: crypto.KeyObject,
  data: Uint8Array,
  signature: Uint8Array,
  hash: string,
  expectedLength: number,
): boolean {
  if (key.type !== "public") {
    throw new InvalidKeyError("expected public key for ECDSA verification");
  }
  if (signature.length !== expectedLength) return false;
  return crypto.verify(hash, data, { key, dsaEncoding: "ieee-p1363" }, signature);
}

// --- Ed25519 ---

function signEd25519(key: crypto.KeyObject, data: Uint8Array): Uint8Array {
  if (key.type !== "private") {
    throw new InvalidKeyError("expected private key for Ed25519 signing");
  }
  return new Uint8Array(crypto.sign(null, data, key));
}

function verifyEd25519(
  key: crypto.KeyObject,
  data: Uint8Array,
  signature: Uint8Array,
): boolean {
  if (key.type !== "public") {
    throw new InvalidKeyError("expected public key for Ed25519 verification");
  }
  return crypto.verify(null, data, key, signature);
}

File: golang/algorithm.go
package httpsig

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/hmac"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/sha512"
	"errors"
	"fmt"
	"hash"
	"math/big"
)

// Algorithm identifies a signature algorithm per RFC 9421 Section 3.3.
type Algorithm string

const (
	AlgorithmRSAPSSSHA512    Algorithm = "rsa-pss-sha512"
	AlgorithmRSAPSSSHA384    Algorithm = "rsa-pss-sha384"
	AlgorithmRSAPSSSHA256    Algorithm = "rsa-pss-sha256"
	AlgorithmRSAV15SHA256    Algorithm = "rsa-v1_5-sha256"
	AlgorithmECDSAP256SHA256 Algorithm = "ecdsa-p256-sha256"
	AlgorithmECDSAP384SHA384 Algorithm = "ecdsa-p384-sha384"
	AlgorithmECDSAP521SHA512 Algorithm = "ecdsa-p521-sha512"
	AlgorithmEd25519         Algorithm = "ed25519"
	AlgorithmHMACSHA256      Algorithm = "hmac-sha256"
	AlgorithmHMACSHA384      Algorithm = "hmac-sha384"
	AlgorithmHMACSHA512      Algorithm = "hmac-sha512"
)

// algorithmSigner dispatches signing to the right algorithm.
func algorithmSign(alg Algorithm, key crypto.Signer, data []byte) ([]byte, error) {
	switch alg {
	case AlgorithmRSAPSSSHA512:
		return signRSAPSS(key, crypto.SHA512, 64, data)
	case AlgorithmRSAPSSSHA384:
		return signRSAPSS(key, crypto.SHA384, 48, data)
	case AlgorithmRSAPSSSHA256:
		return signRSAPSS(key, crypto.SHA256, 32, data)
	case AlgorithmRSAV15SHA256:
		return signRSAV15(key, crypto.SHA256, data)
	case AlgorithmECDSAP256SHA256:
		return signECDSA(key, elliptic.P256(), crypto.SHA256, 32, data)
	case AlgorithmECDSAP384SHA384:
		return signECDSA(key, elliptic.P384(), crypto.SHA384, 48, data)
	case AlgorithmECDSAP521SHA512:
		return signECDSA(key, elliptic.P521(), crypto.SHA512, 66, data)
	case AlgorithmEd25519:
		return signEd25519(key, data)
	default:
		return nil, fmt.Errorf("%w: %s", ErrUnknownAlgorithm, alg)
	}
}

func algorithmVerify(alg Algorithm, pub crypto.PublicKey, data, sig []byte) (bool, error) {
	switch alg {
	case AlgorithmRSAPSSSHA512:
		return verifyRSAPSS(pub, crypto.SHA512, 64, data, sig)
	case AlgorithmRSAPSSSHA384:
		return verifyRSAPSS(pub, crypto.SHA384, 48, data, sig)
	case AlgorithmRSAPSSSHA256:
		return verifyRSAPSS(pub, crypto.SHA256, 32, data, sig)
	case AlgorithmRSAV15SHA256:
		return verifyRSAV15(pub, crypto.SHA256, data, sig)
	case AlgorithmECDSAP256SHA256:
		return verifyECDSA(pub, elliptic.P256(), crypto.SHA256, 32, data, sig)
	case AlgorithmECDSAP384SHA384:
		return verifyECDSA(pub, elliptic.P384(), crypto.SHA384, 48, data, sig)
	case AlgorithmECDSAP521SHA512:
		return verifyECDSA(pub, elliptic.P521(), crypto.SHA512, 66, data, sig)
	case AlgorithmEd25519:
		return verifyEd25519(pub, data, sig)
	default:
		return false, fmt.Errorf("%w: %s", ErrUnknownAlgorithm, alg)
	}
}

// --- RSA-PSS ---

func signRSAPSS(key crypto.Signer, hash crypto.Hash, saltLen int, data []byte) ([]byte, error) {
	rsaKey, ok := key.(*rsa.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("%w: expected *rsa.PrivateKey", ErrInvalidKey)
	}
	h := hash.New()
	h.Write(data)
	digest := h.Sum(nil)
	return rsa.SignPSS(rand.Reader, rsaKey, hash, digest, &rsa.PSSOptions{
		SaltLength: saltLen,
		Hash:       hash,
	})
}

func verifyRSAPSS(pub crypto.PublicKey, hash crypto.Hash, saltLen int, data, sig []byte) (bool, error) {
	rsaPub, ok := pub.(*rsa.PublicKey)
	if !ok {
		return false, fmt.Errorf("%w: expected *rsa.PublicKey", ErrInvalidKey)
	}
	h := hash.New()
	h.Write(data)
	digest := h.Sum(nil)
	err := rsa.VerifyPSS(rsaPub, hash, digest, sig, &rsa.PSSOptions{
		SaltLength: saltLen,
		Hash:       hash,
	})
	if err != nil {
		return false, nil
	}
	return true, nil
}

// --- RSA PKCS1v1.5 ---

func signRSAV15(key crypto.Signer, hash crypto.Hash, data []byte) ([]byte, error) {
	rsaKey, ok := key.(*rsa.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("%w: expected *rsa.PrivateKey", ErrInvalidKey)
	}
	h := hash.New()
	h.Write(data)
	digest := h.Sum(nil)
	return rsa.SignPKCS1v15(rand.Reader, rsaKey, hash, digest)
}

func verifyRSAV15(pub crypto.PublicKey, hash crypto.Hash, data, sig []byte) (bool, error) {
	rsaPub, ok := pub.(*rsa.PublicKey)
	if !ok {
		return false, fmt.Errorf("%w: expected *rsa.PublicKey", ErrInvalidKey)
	}
	h := hash.New()
	h.Write(data)
	digest := h.Sum(nil)
	err := rsa.VerifyPKCS1v15(rsaPub, hash, digest, sig)
	if err != nil {
		return false, nil
	}
	return true, nil
}

// --- ECDSA ---

func signECDSA(key crypto.Signer, curve elliptic.Curve, hash crypto.Hash, fieldLen int, data []byte) ([]byte, error) {
	ecKey, ok := key.(*ecdsa.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("%w: expected *ecdsa.PrivateKey", ErrInvalidKey)
	}
	if ecKey.Curve != curve {
		return nil, fmt.Errorf("%w: expected %s curve", ErrInvalidKey, curve.Params().Name)
	}
	h := hash.New()
	h.Write(data)
	digest := h.Sum(nil)
	r, s, err := ecdsa.Sign(rand.Reader, ecKey, digest)
	if err != nil {
		return nil, err
	}
	// RFC 9421: signature is r || s, each fieldLen bytes, big-endian, zero-padded
	rBytes := r.Bytes()
	sBytes := s.Bytes()
	sigLen := fieldLen * 2
	sig := make([]byte, sigLen)
	copy(sig[fieldLen-len(rBytes):fieldLen], rBytes)
	copy(sig[sigLen-len(sBytes):sigLen], sBytes)
	return sig, nil
}

func verifyECDSA(pub crypto.PublicKey, curve elliptic.Curve, hash crypto.Hash, fieldLen int, data, sig []byte) (bool, error) {
	ecPub, ok := pub.(*ecdsa.PublicKey)
	if !ok {
		return false, fmt.Errorf("%w: expected *ecdsa.PublicKey", ErrInvalidKey)
	}
	if ecPub.Curve != curve {
		return false, fmt.Errorf("%w: expected %s curve", ErrInvalidKey, curve.Params().Name)
	}
	sigLen := fieldLen * 2
	if len(sig) != sigLen {
		return false, nil
	}
	r := new(big.Int).SetBytes(sig[:fieldLen])
	s := new(big.Int).SetBytes(sig[fieldLen:])
	h := hash.New()
	h.Write(data)
	digest := h.Sum(nil)
	return ecdsa.Verify(ecPub, digest, r, s), nil
}

// --- Ed25519 ---

func signEd25519(key crypto.Signer, data []byte) ([]byte, error) {
	edKey, ok := key.(ed25519.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("%w: expected ed25519.PrivateKey", ErrInvalidKey)
	}
	return ed25519.Sign(edKey, data), nil
}

func verifyEd25519(pub crypto.PublicKey, data, sig []byte) (bool, error) {
	edPub, ok := pub.(ed25519.PublicKey)
	if !ok {
		return false, fmt.Errorf("%w: expected ed25519.PublicKey", ErrInvalidKey)
	}
	return ed25519.Verify(edPub, data, sig), nil
}

// --- HMAC ---

func signHMACSHA256(secret []byte, data []byte) ([]byte, error) {
	return signHMAC(secret, sha256.New, data)
}

func verifyHMACSHA256(secret []byte, data, sig []byte) (bool, error) {
	return verifyHMAC(secret, sha256.New, data, sig)
}

func signHMACSHA384(secret []byte, data []byte) ([]byte, error) {
	return signHMAC(secret, sha512.New384, data)
}

func verifyHMACSHA384(secret []byte, data, sig []byte) (bool, error) {
	return verifyHMAC(secret, sha512.New384, data, sig)
}

func signHMACSHA512(secret []byte, data []byte) ([]byte, error) {
	return signHMAC(secret, sha512.New, data)
}

func verifyHMACSHA512(secret []byte, data, sig []byte) (bool, error) {
	return verifyHMAC(secret, sha512.New, data, sig)
}

func signHMAC(secret []byte, hashFunc func() hash.Hash, data []byte) ([]byte, error) {
	if len(secret) == 0 {
		return nil, errors.New("httpsig: empty HMAC secret")
	}
	mac := hmac.New(hashFunc, secret)
	mac.Write(data)
	return mac.Sum(nil), nil
}

func verifyHMAC(secret []byte, hashFunc func() hash.Hash, data, sig []byte) (bool, error) {
	expected, err := signHMAC(secret, hashFunc, data)
	if err != nil {
		return false, err
	}
	return hmac.Equal(sig, expected), nil
}
